pipelined/bevy_pbr2/src/material.rs

```diff
@@ -134,26 +138,59 @@ impl Plugin for StandardMaterialPlugin {
 #[derive(Debug, Clone)]
 pub struct GpuStandardMaterial {
     pub buffer: Buffer,
-    // FIXME: image handles feel unnecessary here but the extracted asset is discarded
-    pub base_color_texture: Option<Handle<Image>>,
-    pub emissive_texture: Option<Handle<Image>>,
-    pub metallic_roughness_texture: Option<Handle<Image>>,
-    pub occlusion_texture: Option<Handle<Image>>,
+    pub bind_group: BindGroup,
 }
 
 impl RenderAsset for StandardMaterial {
     type ExtractedAsset = StandardMaterial;
     type PreparedAsset = GpuStandardMaterial;
+    type Param = (
+        SRes<RenderDevice>,
+        SRes<PbrShaders>,
+        SRes<RenderAssets<Image>>,
+    );
 
     fn extract_asset(&self) -> Self::ExtractedAsset {
         self.clone()
     }
 
     fn prepare_asset(
         material: Self::ExtractedAsset,
-        render_device: &RenderDevice,
-        _render_queue: &RenderQueue,
-    ) -> Self::PreparedAsset {
+        (render_device, pbr_shaders, gpu_images): &mut SystemParamItem<Self::Param>,
+    ) -> Result<Self::PreparedAsset, PrepareAssetError<Self::ExtractedAsset>> {
+        let (base_color_texture_view, base_color_sampler) = if let Some(result) =
```

Comment:
All those if let could be rewritten with ok_or to simplify it a bit
```rust
let (base_color_texture_view, base_color_sampler) = 
	image_handle_to_view_sampler(pbr_shaders, gpu_images, &material.base_color_texture)
		.ok_or(PrepareAssetError::RetryNextUpdate(material))?;
```

Reply:
Actually they cant! I'd love to, but RetryNextUpdate(material) takes ownership of `material`, which causes subsequent lines to fail. By including the "branching" inline, rust can reason about the control flow / only moves `material` when it is actually needed.